src/components/sets/io.js

```diff
@@ -0,0 +1,204 @@
+import { dsvFormat } from 'd3-dsv';
+import Ajv from 'ajv';
+import tinycolor from 'tinycolor2';
+import { parse as json2csv } from 'json2csv';
+import { version } from '../../../package.json';
+import { PATH_SEP } from './sets';
+
+import hierarchicalSetsSchema from '../../schemas/hierarchical-sets.schema.json';
+
+export const tabularFileType = 'TSV';
+export const tabularFileExtension = 'tsv';
+const tabularColumnSeparator = '\t';
+const tabularHierarchySeparator = ';';
+const tabularNA = 'NA';
+
+/**
+ * Check whether the elements of two arrays are equal.
+ * @param {Array} a One of the two arrays.
+ * @param {Array} b The other of the two arrays.
+ * @returns {boolean} Whether the two arrays contain the same elements.
+ */
+function arraysEqual(a, b) {
+  return a.length === b.length && a.every((v, i) => v === b[i]);
+}
+
+/**
+   * Make a timestamped name for an import root node.
+   * @returns {string} A new name for an import node.
+   */
+function makeImportName() {
+  const timestamp = (new Date()).toLocaleString();
+  return `Import ${timestamp}`;
+}
+
+/**
+   * Handler for tabular (TSV) imports.
+   * @param {object} props The component props.
+   * @param {string} result The data passed from the onImport function as a string.
+   */
+export function handleImportTabular(props, result) {
+  const {
+    setsTree,
+  } = props;
+  const dsvParser = dsvFormat(tabularColumnSeparator);
+  /**
+     * Convert a string color representation to an array of [r,g,b].
+     * @param {string} colorString The color as a string.
+     * @returns {Array} The color as an array.
+     */
+  function colorAsArray(colorString) {
+    return Object.entries(tinycolor(colorString).toRgb())
+      .filter(c => c[0] !== 'a').map(c => c[1]);
+  }
+  // Fall back to set name if set key not provided.
+  const importData = dsvParser.parse(result, row => ({
+    itemId: row['Item ID'],
+    setName: row['Set Name'],
+    setKey: row['Set Key'] || row['Set Name'].split(tabularHierarchySeparator).join(PATH_SEP),
+    setColor: colorAsArray(row['Set Color']),
+  }));
+    // Make array of unique set key strings.
+  const importedSetKeys = Array.from(new Set(importData.map(d => d.setKey)));
+  // Construct the array representation of the tree required by the tree import function.
+  const importedSetsTree = [];
+  // Iterate over each set and append to the tree array representation.
+  importedSetKeys.forEach((setKey) => {
+    const setItems = importData.filter(d => d.setKey === setKey);
+    const isEmpty = setItems.length === 1 && setItems[0].itemId === tabularNA;
```

Comment:
Because this is a tabular format where each row is an item in a set, I am defining a node in the hierarchy that has no items associated as one row with the value "NA" as the Item ID. Otherwise the set would never appear in the exported file and not be able to be imported.